### PR TITLE
[Draft] Cleaner backoff utilities in Beam

### PR DESCRIPTION
**Pull request by dhalperi**
We have 3 different backoff classes, which don't really have that much different functionality. Add a single, flexible backoff implementation that can be used to replace all three classes. Additionally, this new backoff actually supports more functionality than any of the other three did -- you can limit retries, cap the exponential growth of an individual backoff, and cap the cumulative time spent in backoff; prior implementations did not allow all 3.

This also makes the parameters self-obvious (`Duration`, not number-that-is-also-millis) where appropriate.

This initial PR should have no functional changes.

TODO:
- [x] is there a better name than "FlexibleBackoff"
- [x] javadoc on FlexibleBackof

?

### CODE REVIEW

**Comment by dhalperi** [comment]
R: @bjchambers @peihe

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java [hunk 1]
```diff
@@ -250,7 +249,12 @@ State waitUntilFinish(
       }
 
       if (!hasError) {
+        // Reset the backoff, and update the new cumulative sleep time to be the remaining
+        // part of the total input sleep duration.
         backoff.reset();
+        long nanosConsumed = nanoClock.nanoTime() - startNanos;
+        Duration consumed = Duration.millis((nanosConsumed + 999999) / 1000000);
+        backoff = backoff.withMaxCumulativeBackoff(duration.minus(consumed));
```
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java]: Could we track the cumulative time inside the FlexibleBackoff, and have a getCumulativeTime() or getRemainingTime()?

I think that will be more cleaner on the client side, and it also can be reused.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+  public FlexibleBackoff withExponent(double exponent) {
+    checkArgument(exponent > 0, "exponent %s must be greater than 0", exponent);
+    return new FlexibleBackoff(
+        exponent, initialBackoff, maxBackoff, maxCumulativeBackoff, maxAttempts);
+  }
+
+  public FlexibleBackoff withInitialBackoff(Duration initialBackoff) {
+    checkArgument(
+        initialBackoff.getMillis() > 0,
```
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java]: be consistent with "checkArgument(maxCumulativeBackoff.isLongerThan(Duration.ZERO))"?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+  }
+
+  @Override
+  public long nextBackOffMillis() {
+    // Maximum number of attempts reached.
+    if (currentAttempt >= maxAttempts) {
+      return BackOff.STOP;
+    }
+    // Maximum cumulative backoff reached.
+    if (currentCumulativeBackoff.equals(maxCumulativeBackoff)) {
```
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java]: isLongerThan? and add tests

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+      return BackOff.STOP;
+    }
+
+    double currentIntervalMillis =
+        Math.min(
+            initialBackoff.getMillis() * Math.pow(exponent, currentAttempt),
```
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java]: We can keep track of currentBackoffFactor, so it doesn't have to do pow() every time.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java (hunk 3 above)
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java]: I see we have a cap limit on cumulative backoff.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java (hunk 1 above)
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java]: I thought about then when replacing the other class, but I think that it is a bad break of the BackOff abstraction. For example:

- `reset()` only doing a partial reset (attempts, not time)
- definition of start time (is it when the BackOff is constructed? when the first call to `nextBackOffMillis()`? Something else?

It seemed unnecessary complexity for the `BackOff` object itself. Callers should be the ones who care about this and worry about the details. It's not that bad and it keeps the purpose of `BackOff` as it is documented today.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java (hunk 2 above)
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java (hunk 4 above)
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java]: unnecessary optimization -- this is not perf critical code

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java (hunk 3 above)
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java]: Tests are there: `testBoundedIntervalAndCumTimeWithReset`

Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java]: switched to `.compareTo() >= 0` to mimic `>=` as in the above `if`.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java (hunk 4 above)
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java]: (If github showed branching, this is branched from the `IntervalBounded` code I'm deleting)

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java (hunk 1 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java]: I agree BackOff shouldn't see the cumulative time.
But, here we are using FlexibleBackoff, and update MaxCumulativeBackoff each time.

It will also become easier to test if we put the tracking in FlexibleBackoff.

Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java]: The value tracked by the BackOff is a different number than the code here wants. The code here wants wall-clock consumed.

File: examples/java/src/main/java/org/apache/beam/examples/common/ExampleUtils.java [hunk 5]
```diff
@@ -79,7 +80,9 @@ public ExampleUtils(PipelineOptions options) {
    */
   public void setup() throws IOException {
     Sleeper sleeper = Sleeper.DEFAULT;
-    BackOff backOff = new AttemptBoundedExponentialBackOff(3, 200);
+    BackOff backOff =
+        FlexibleBackoff.of()
```
Comment (bjchambers) [inline examples/java/src/main/java/org/apache/beam/examples/common/ExampleUtils.java]: `FluentBackoff` (as analog to FluentIterable, etc.)

I'd also consider getting rid of the `of()` static factory, and instead just requiring the methods to configure it to be called.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java [hunk 6]
```diff
@@ -214,21 +213,22 @@ State waitUntilFinish(
...
-                MESSAGES_POLLING_INTERVAL,
-                duration.getMillis(),
-                AttemptAndTimeBoundedExponentialBackOff.ResetPolicy.ATTEMPTS,
-                nanoClock)
-            : new AttemptBoundedExponentialBackOff(
-                MESSAGES_POLLING_ATTEMPTS, MESSAGES_POLLING_INTERVAL);
+    FlexibleBackoff backoff = FlexibleBackoff.of()
+        .withMaxRetries(MESSAGES_POLLING_RETRIES)
+        .withInitialBackoff(MESSAGES_POLLING_INTERVAL);
+    if (duration.isLongerThan(Duration.ZERO)) {
```
Comment (bjchambers) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java]: `withMaxCumulativeBackoff` could have the logic that duration <= 0 means no limit. Then the if statement can move into the utility.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java [hunk 7]
```diff
@@ -214,21 +213,22 @@ State waitUntilFinish(
...
+      backoff = backoff.withMaxCumulativeBackoff(duration);
```
Comment (bjchambers) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java]: `withMaxCumulativePerRequestBackoff` (to emphasize this is maximum backoff on a single request)?

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java [hunk 8]
```diff
@@ -214,21 +213,22 @@ State waitUntilFinish(
...
+    }
+
+    // In this function, track the cumulative time since we've started backing off and use this
+    // to enforce a wall-clock limit.
+    long startNanos = nanoClock.nanoTime();
```
Comment (bjchambers) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java]: Why is this necessary?

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java [hunk 9]
```diff
@@ -214,21 +213,22 @@ State waitUntilFinish(
...
+
     State state;
     do {
       // Get the state of the job before listing messages. This ensures we always fetch job
       // messages after the job finishes to ensure we have all them.
-      state = getStateWithRetries(1, sleeper);
+      state = getStateWithRetries(0, sleeper);
```
Comment (bjchambers) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java]: why this change?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+  private Duration currentCumulativeBackoff;
+  private int currentRetry;
+
+  /**
+   * Instantiates a {@link FlexibleBackoff} with the default exponential backoff and initial
+   * backoff.
+   *
+   * @see FlexibleBackoff
+   */
+  public static FlexibleBackoff of() {
```
Comment (bjchambers) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java]: I think this would the right place to document defaults, rather than the class Javadoc.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+ * after exponential increase can be capped using {@link FlexibleBackoff#withMaxBackoff(Duration)}.
+ *
+ * <p>By default, and after modifying any of the above parameters, {@link FlexibleBackoff} does not
+ * limit the number of retries. To limit the backoff, the maximum total number of retries can be set
+ * using {@link FlexibleBackoff#withMaxRetries(int)}. The total time spent in backoff can be
+ * time-bounded as well by configuring {@link FlexibleBackoff#withMaxCumulativeBackoff(Duration)}.
+ * If either of these limits are reached, the {@link FlexibleBackoff#nextBackOffMillis()} will
+ * return {@link BackOff#STOP} to signal that no more retries should continue.
+ */
+public final class FlexibleBackoff implements BackOff {
```
Comment (bjchambers) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java]: I'm wondering if instead, this should be a class for configuring a BackOff pool or factory.

```
public final class FluentBackoff {

  public static final FluentBackoff DEFAULT = ...;

  public BackOff backoff() { ... }
  public withCrossRequestLimit(Duration duration) { ... }
  public withPerRequestLimit(Duration duration) { ... }
  public withPerAttemptLimit(Duration duration) { ... }
  ...
}
```

The benefit is that it allows controlling backoff across multiple requests.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+
+
+/**
+ * A {@link BackOff} for which the retry algorithm is customizable. By default the
+ * {@link FlexibleBackoff} uses exponential backoff (base exponent 1.5) with an initial backoff of
+ * 1 second. These parameters can be overridden with {@link FlexibleBackoff#withExponent(double)}
+ * and {@link FlexibleBackoff#withInitialBackoff(Duration)}, respectively, and the maximum backoff
+ * after exponential increase can be capped using {@link FlexibleBackoff#withMaxBackoff(Duration)}.
+ *
+ * <p>By default, and after modifying any of the above parameters, {@link FlexibleBackoff} does not
```
Comment (bjchambers) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java]: This sounds weird -- each of the modifier methods should document what its default value was (in addition to the default creation method specifying that). Modifying one of the parameters shouldn't modify other parameters (eg., once set, the number of retries shouldn't go back to unlimitedwhen other things are modified).

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+ * limit the number of retries. To limit the backoff, the maximum total number of retries can be set
+ * using {@link FlexibleBackoff#withMaxRetries(int)}. The total time spent in backoff can be
+ * time-bounded as well by configuring {@link FlexibleBackoff#withMaxCumulativeBackoff(Duration)}.
+ * If either of these limits are reached, the {@link FlexibleBackoff#nextBackOffMillis()} will
+ * return {@link BackOff#STOP} to signal that no more retries should continue.
+ */
+public final class FlexibleBackoff implements BackOff {
+  private static final double DEFAULT_EXPONENT = 1.5;
+  private static final double DEFAULT_RANDOMIZATION_FACTOR = 0.5;
+  private static final Duration DEFAULT_MIN_BACKOFF = Duration.standardSeconds(1);
```
Comment (bjchambers) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java]: I think it would clarify if these were `DEFAULT_MIN_ATTEMPT_BACKOFF` and `DEFAULT_MAX_ATTEMPT_BACKOFF`. Then "cumulative" is just `DEFAULT_MAX_BACKOFF`, I believe?

Specifically, "request" = a specific request we'd like to make, "attempt" = an attempt to issue that request.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+  private static final int DEFAULT_MAX_RETRIES = Integer.MAX_VALUE;
+  private static final Duration DEFAULT_MAX_CUM_BACKOFF = Duration.standardDays(1000);
+  // Customization of this backoff.
+  private final double exponent;
+  private final Duration initialBackoff;
+  private final Duration maxBackoff;
+  private final Duration maxCumulativeBackoff;
+  private final int maxRetries;
+  // Current state
+  private Duration currentCumulativeBackoff;
```
Comment (bjchambers) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java]: As discussed -- suggest the actual Backoff be created (via a `createBackoff` or `build()` method), which allows the mutable details of a backoff to go into the built instance.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+ */
+public final class FlexibleBackoff implements BackOff {
+  private static final double DEFAULT_EXPONENT = 1.5;
+  private static final double DEFAULT_RANDOMIZATION_FACTOR = 0.5;
+  private static final Duration DEFAULT_MIN_BACKOFF = Duration.standardSeconds(1);
+  private static final Duration DEFAULT_MAX_BACKOFF = Duration.standardDays(1000);
+  private static final int DEFAULT_MAX_RETRIES = Integer.MAX_VALUE;
+  private static final Duration DEFAULT_MAX_CUM_BACKOFF = Duration.standardDays(1000);
+  // Customization of this backoff.
+  private final double exponent;
```
Comment (bjchambers) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java]: `@AutoValue` for the final configuration bits?

Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java]: discussed in person, not for now.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java (hunk 14 above)
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java (hunk 12 above)
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java]: PTAL

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java (hunk 11 above)
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java]: did the configuration, did not handle multiple requests as a deliberate choice. Following the Google API libraries (which set the BackOff interface) here.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java (hunk 10 above)
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java]: done

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java (hunk 9 above)
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java]: The 3 custom BackOff implementations we had were very casual and inconsistent about "retries" vs "attempts". This function intended for there to be exactly 1 attempt, but said 1 retry. Hence this change (and several similar).

File: examples/java/src/main/java/org/apache/beam/examples/common/ExampleUtils.java (hunk 5 above)
Comment (dhalperi) [inline examples/java/src/main/java/org/apache/beam/examples/common/ExampleUtils.java]: done

**Comment by dhalperi** [comment]
PTAL

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java [hunk 16]
```diff
@@ -95,14 +93,27 @@
...
    */
-  static final int MESSAGES_POLLING_ATTEMPTS = 12;
-  static final int STATUS_POLLING_ATTEMPTS = 5;
+  static final int MESSAGES_POLLING_RETRIES = 11;
+  static final int STATUS_POLLING_RETRIES = 4;
+
+  private static final FluentBackoff MESSAGES_BACKOFF_FACTORY =
+      FluentBackoff.DEFAULT
+          .withInitialBackoff(MESSAGES_POLLING_INTERVAL)
+          .withMaxRetries(MESSAGES_POLLING_RETRIES)
```
Comment (bjchambers) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java]: nit: We could have `withMaxAttempts` which simplify things a little (wouldn't need to change the name of the statics and may be clearer in some cases (if we want to limit the number of attempts). But this seems fine.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java (hunk 8 above)
Comment (bjchambers) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java]: Maybe clarify the comment to state that the backoff utility doesn't allow for limiting time from first-attempt to success? Should it?

Or is this time from first-attempt at first request to success of all requests (where each request is separately retried)?

Comment (bjchambers) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java]: Ah, looking at more of the context, yeah. This comment should be:

```
// This function tracks the cumulative time from the *first request* to enforce the wall-clock limit.
// Any backoff instance could, at best, track the the time since the first attempt at a given request.
// Thus, we need to track the cumulative time ourselves.
```

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java [hunk 17]
```diff
@@ -250,7 +262,16 @@ State waitUntilFinish(
       }
 
       if (!hasError) {
+        // Reset the backoff.
         backoff.reset();
```
Comment (bjchambers) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java]: Rather than resetting, would it make sense to just *always* recreate the backoff?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/FluentBackoff.java [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+  /**
+   * By default the {@link BackOff} created by this builder will use exponential backoff (base
+   * exponent 1.5) with an initial backoff of 1 second. These parameters can be overridden with
+   * {@link #withExponent(double)} and {@link #withInitialBackoff(Duration)},
+   * respectively, and the maximum backoff after exponential increase can be capped using {@link
+   * FluentBackoff#withMaxBackoff(Duration)}.
+   *
+   * <p>The default {@link BackOff} does not limit the number of retries. To limit the backoff, the
+   * maximum total number of retries can be set using {@link #withMaxRetries(int)}. The
+   * total time spent in backoff can be time-bounded as well by configuring {@link
```
Comment (bjchambers) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/FluentBackoff.java]: Is this the "total time spent in backoff" or "total time spent attempting this request"?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/FluentBackoff.java [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+   */
+  public FluentBackoff withMaxRetries(int maxRetries) {
+    checkArgument(maxRetries >= 0, "maxRetries %s cannot be negative", maxRetries);
+    return new FluentBackoff(
+        exponent, initialBackoff, maxBackoff, maxCumulativeBackoff, maxRetries);
+  }
+
+  private static class BackoffImpl implements BackOff {
+
+    // Customization of this backoff.
```
Comment (bjchambers) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/FluentBackoff.java]: Since `FluentBackoff` is immutable, this could just hold on to a pointer to the `FluentBackoff` that created it to get the configuration values from thath.

**Comment by bjchambers** [comment]
Few nits, but LGTM

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/FluentBackoff.java (hunk 19 above)
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/FluentBackoff.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/FluentBackoff.java [hunk 20]
```diff
@@ -150,14 +149,20 @@ public FluentBackoff withMaxRetries(int maxRetries) {
         exponent, initialBackoff, maxBackoff, maxCumulativeBackoff, maxRetries);
   }
 
+  public String toString() {
+    return MoreObjects.toStringHelper(FluentBackoff.class)
```
Comment (bjchambers) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/FluentBackoff.java]: I wonder if it would be nice to include a line that is essentially "backoffs" (and just run out the maxRetries of the exponent). And/or provide a utility for that. And/or indicate what the total time spent backing off before failure would be (eg., when maxCumulativeBackoff isn't set, we would need to add up all the backoffs across the attempts).

File: sdks/java/core/src/test/java/org/apache/beam/sdk/util/FluentBackoffTest.java [hunk 21]
```diff
@@ -188,19 +188,31 @@ public void testBoundedIntervalAndCumTimeAndRetriesWithReset() throws Exception
...
+    FluentBackoff config = FluentBackoff.DEFAULT
+        .withExponent(3.4)
+        .withMaxRetries(4)
+        .withInitialBackoff(Duration.standardSeconds(3))
+        .withMaxBackoff(Duration.standardHours(1))
+        .withMaxCumulativeBackoff(Duration.standardDays(1));
+    BackOff backOff = config.backoff();
+
+    assertEquals(
+        "BackoffImpl{backoffConfig=" + config.toString() + ","
```
Comment (bjchambers) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/util/FluentBackoffTest.java]: +1 I especially like that this separates the "here are the static configuration from current state" even in the toString!

**Comment by bjchambers** [comment]
LGTM
